### ISSUE TITLE
Bug when multiple entries?

### ISSUE THREAD

**Issue by smurfb** [kotope/aio_energy_management#18]
Thanks for your hard work, i tried to expand to a couple of more entries to get everything to run on the cheapest periods. 

But it seems like if you have have many sensors that have the same timeframe (but different numbers of hours) and they are supposed to start on the same time, only one will start. 

Note that they all show up in the calander, its just the sensor that not switches to on. 

From config: 
  cheapest_hours:
    - nordpool_entity: sensor.nordpool_kwh_se3_sek_3_10_025
      unique_id: cheapesthours_2
      name: Cheapest hours 2
      first_hour: 0
      last_hour: 23
      starting_today: false
      number_of_hours: 2
      sequential: false
      failsafe_starting_hour: 0
    - nordpool_entity: sensor.nordpool_kwh_se3_sek_3_10_025
      unique_id: cheapesthours_3
      name: Cheapest hours 3
      first_hour: 0
      last_hour: 23
      starting_today: false
      number_of_hours: 3
      sequential: false
      failsafe_starting_hour: 0
    - nordpool_entity: sensor.nordpool_kwh_se3_sek_3_10_025
      unique_id: cheapesthours_5
      name: Cheapest hours 5
      first_hour: 0
      last_hour: 23
      starting_today: false
      number_of_hours: 5
      sequential: false
      failsafe_starting_hour: 0

![Screenshot_20241005_030427_Home Assistant](https://github.com/user-attachments/assets/4b0e31f2-06a5-4da2-9492-2c69d183f70b)
![Screenshot_20241005_030441_Home Assistant](https://github.com/user-attachments/assets/4a54b6be-e4fa-44b0-b03c-d0f48e32fbe0)

Thanks again for your work!

**Comment by kotope** [comment]
This sounds odd as separate sensors only share the persistent storage. They all are run on a separate update loop and should not interfere with each other.

Anyhow, I'll have a look on this soon and see if I can reproduce the issue.

**Comment by kotope** [comment]
A couple of questions though:
 - Did you reboot or reload integration between yesterday and today?
 - Are you running latest 0.2.2 version?

**Comment by smurfb** [comment]
Yes running the latest.  

And yes I think I gave rebooted HA.

**Comment by kotope** [comment]
I'll made a similar setup on my dev env and see if I can get this to happen also as I couldn't see any reason from the code why this would happen.

**Comment by kotope** [comment]
@smurfb one more thing.. could you also paste a screenshot of your working sensor attributes?

**Comment by kotope** [comment]
If the state is still off, could you send me debug logs of it? Most probably I could figure the problem out of it.

Go to integrations -> aio energy management -> enable debug logging and allow the system run for few minutes.
After that go to same place and disable logging. It should download the component logs on stop.

You can attach the logs here or send by email to creatingsmarthome@gmail.com 

![Screenshot_20241005-160444](https://github.com/user-attachments/assets/19ae35c5-d631-41b4-8d8f-74d349abb0e5)

**Comment by smurfb** [comment]
Here is the debug log file, I will leave it running so that it will be on debug when the  clock strikes 12 and cheapest hours start. 

I tried to look in the logfile but it is filled with rubbish from a tuya zigbee device that is spamming errors. But I have removed that so I can have a look in the logfile. 

This night the prices were too high so it was only the 5h sensor that was supposed to start at 3 (but it didn't).  But all three (2h, 3h, 5h) should start at noon. So I will keep the log going and hopefully there will be something. 

The calendar is always working though! 

[aio_ha.log](https://github.com/user-attachments/files/17268479/aio_ha.log)

**Comment by kotope** [comment]
Thanks! As the logs are from today, can you also send the attributes that "Cheapest hours 5" is now displaying? 

The issue here is that the "list_next" conversion to "list" is not made properly. After 'list' has expired the swap should be made on next update loop (called by Home Assistant core every 30s). Even if it fails it should retry on next iteration.

**Comment by smurfb** [comment]
Thanks for explaining!

Here is the current cheapest hour 5. 

![Screenshot_20241005_030441_Home Assistant (1)](https://github.com/user-attachments/assets/e58e8f36-afd5-46cf-9402-1eed6d140957)

**Comment by smurfb** [comment]
Thanks for explaining!

Here is the current cheapest hour 5 attributes. 

<img width="558" alt="Skärmavbild 2024-10-06 kl  09 32 19" src="https://github.com/user-attachments/assets/f048978d-3e97-42b8-ab8e-2b3a36f6af74">

**Comment by kotope** [comment]
Looks like that data has been updated properly now and most probably the sensor will work ok today.

If you encounter the similar issue in the future, please send few minutes of AIO Energy Management debug logs and the sensor attributes. :-)

I'm also running similar configurations on my dev env now as well to see if this happens for me as well.

**Comment by smurfb** [comment]
Yes, it seemed to work now! So the problem is somehow connected me restarting the machine?

Thanks for all your help!

**Comment by teelahti** [comment]
If it helps: I had a similar situation last night even though I have only one sensor with very basic configuration:

```YAML
aio_energy_management:
    cheapest_hours:
      - nordpool_entity: sensor.nordpool_kwh_fi_eur_3_095_024
        unique_id: cheapest_electricity_hours
        name: Cheapest electricity hours
        first_hour: 00
        last_hour: 23
        starting_today: false
        number_of_hours: 8
        sequential: false
        failsafe_starting_hour: 0
    calendar:
      name: Energy Management
      unique_id: energy_management_calendar
```

Calendar showed cheapest hours 23-00, and 02-06, but the binary sensor never turned on at 02.00.

**Comment by kotope** [comment]
Thanks @teelahti .

I'm still trying to get this reproduced, but so far no luck. 
Could you try to get the AIO Energy Management logs as described above and paste current device attributes? If I can get get the logs from in error situation happen I might be able to find the root cause of this. :-)

And did you reboot your HA yesterday or has it been running more than a day?

**Comment by teelahti** [comment]
> And did you reboot your HA yesterday or has it been running more than a day?

I _did_ boot yesterday after setting everything up (moved from previous script based approach to AIO). So the root cause might very well be the same. 

Attributes: 

<img width="534" alt="image" src="https://github.com/user-attachments/assets/1b0fc38e-8cfa-4624-bae8-abb20fd1845a">

I don't have debug logging enabled, but will put it on if this continues. Btw. thanks for super useful addition for HA. this is exactly what I need (especially if you plan to add some hard price limits and such).

**Comment by kotope** [comment]
Ok, thanks @teelahti . Looks like attributes are ok now.

I think I have a hunch now where the problem might lie, most probably list_next date conversion goes wrong on persistent storage loading after a reboot. I'll write few unit tests this evening to see if I find the issue in there.

**Comment by kotope** [comment]
Found it. Will fix this and release on 0.3.0 later this week.

Was related to the loading from persistent storage after a reboot. list_next expiration date conversion was not made properly.

**Comment by smurfb** [comment]
Great news for us that reboots the machine too often. Thanks for your persistance.

**Comment by kotope** [comment]
Fixed with 0.3.0. Please re-open if issue still persists :)
